vault/backend.go

```diff
@@ -0,0 +1,616 @@
+			}
+		}
+
+		strPurpose := fmt.Sprintf("%d", purpose)
+		strCoin := fmt.Sprintf("%d", coin)
+		strAct := fmt.Sprintf("%d", act)
+
+		strListing += "        {\n"
+
+		strListing += "            \"name\": \""
+		if act == 0 {
+			strListing += "default"
+		} else {
+			strListing += "act:" + strAct
+		}
+		strListing += "\",\n"
+
+		strListing += "            \"address_type\": \"" + addrType +
+			"\",\n"
+
+		strListing += "            \"extended_public_key\": \"" +
+			xPub.String() + "\",\n"
+
+		strListing += "            \"master_key_fingerprint\": null,\n"
+
+		strListing += "            \"derivation_path\": \"m/" +
+			strPurpose + "'/" + strCoin + "'/" + strAct + "'\",\n"
+
+		strListing += "            \"external_key_count\": 0,\n"
+
+		strListing += "            \"internal_key_count\": 0,\n"
+
+		strListing += "            \"watch_only\": false\n"
+
+		strListing += "        }"
+
+		return strListing, nil
+	}
+
+	for _, acctInfo := range defaultPurposes {
+		strListing, err := listAccount(
+			acctInfo.purpose,
+			0,
+			0,
+			acctInfo.addrType,
+			acctInfo.hdVersion[net.HDCoinType][:],
+		)
+		if err != nil {
+			b.Logger().Error("Failed to derive default account",
+				"node", strNode, "err", err)
+			return nil, err
+		}
+
+		acctList += strListing + ",\n"
+	}
+
+	for act := uint32(0); act <= MaxAcctID; act++ {
+		strListing, err := listAccount(
+			Bip0043purpose,
+			net.HDCoinType,
+			act,
+			"WITNESS_PUBKEY_HASH",
+			nil,
+		)
+		if err != nil {
+			b.Logger().Error("Failed to derive Lightning account",
+				"node", strNode, "err", err)
+			return nil, err
+		}
+
+		acctList += strListing
+
+		if act < MaxAcctID {
+			acctList += ","
+		}
+
+		acctList += "\n"
+	}
+
+	acctList += "    ]\n}"
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"acctList": acctList,
+		},
+	}, nil
+}
+
+func (b *backend) ecdh(ctx context.Context, req *logical.Request,
+	data *framework.FieldData) (*logical.Response, error) {
+
+	peerPubHex := data.Get("peer").(string)
+	if len(peerPubHex) != 2*btcec.PubKeyBytesLenCompressed {
+		b.Logger().Error("Peer pubkey is wrong length",
+			"peer", peerPubHex)
+		return nil, errors.New("invalid peer pubkey")
+	}
+
+	peerPubBytes, err := hex.DecodeString(peerPubHex)
+	if err != nil {
+		b.Logger().Error("Failed to decode peer pubkey hex",
+			"error", err)
+		return nil, err
+	}
+
+	peerPubKey, err := btcec.ParsePubKey(peerPubBytes)
+	if err != nil {
+		b.Logger().Error("Failed to parse peer pubkey",
+			"error", err)
+		return nil, err
+	}
+
+	var (
+		pubJacobian btcec.JacobianPoint
+		s           btcec.JacobianPoint
+	)
+	peerPubKey.AsJacobian(&pubJacobian)
+
+	strNode := data.Get("node").(string)
+
+	seed, net, err := b.getNode(ctx, req.Storage, strNode)
+	if err != nil {
+		b.Logger().Error("Failed to retrieve node info",
+			"node", strNode, "error", err)
+		return nil, err
+	}
+	defer zero(seed)
+
+	privKey, err := derivePrivKey(seed, net, data.Get("path").([]int))
```

Comment:
I see that in lnd, there is also some scanning going on: https://github.com/lightningnetwork/lnd/blob/71fc52dfe38425723fe40dbf16af58ada151cb76/keychain/btcwallet.go#L342

Does that need to be mirrored in lndsigner?

Reply:
I haven't seen any evidence that this scanning is used when using the remote signer. I can't find evidence it's used at all outside tests, but haven't looked closely enough. If we do need to scan, we can probably just do it with the xpubs in the sidecar instead of in `vault`?

Reply:
If we run into the case that we should be running the scanning code to find the private key, do we then just return an error? I think so, but asking to be sure.

Reply:
Yes, we'll get an error. The scanning above happens when the index (last element) in the derivation path is 0 *and* there's a specified pubkey. When that's passed to the vault plugin, it will derive the pubkey at the specified index (0) and check against the specified pubkey; if it doesn't match, it will return a pubkey mismatch error immediately after: https://github.com/bottlepay/lndsigner/blob/master/vault/backend.go#L225

Reply:
Okay, we should be good then.